# Workshop - Data Science & Data Mining
###### NEREUS-USP, São Paulo, April 22, 2019


### *Leonardo Caio de Ladalardo Martins*
lcladalardo@gmail.com

_*Undergraduate student at FEA-USP_

## Section 5: Data Mining - Applications

This case study focuses on the CNPJ Public Database, available at [Receita Federal](http://receita.economia.gov.br/orientacao/tributaria/cadastros/cadastro-nacional-de-pessoas-juridicas-cnpj/dados-publicos-cnpj). This database is very rich in terms of details, but really heavy as well, it contains about:
1. 5gb of data (compressed) and almost 90gb of pure data (not compressed);
2. More than 140mn of rows of diferent layout tables;
3. Personal information about all firms (PJ) and partners (PF) in Brazil;
4. An impressive amount of geographic information about firms (but requiring to be cleaned as well);
5. Almost 70mn of entries overall formats;

In [2]:
#Loading important libraries
%matplotlib inline
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sb
import time 
import numpy as np
import glob

## Section 5.1: A Case Study - Reading the CNPJ Database

The aim of this section is to provide an example how to load real data. As we're going to see, 'Real Data is messy', but we need to create a strategy to deal with the potential issues that are going to arrive. 

#### What we know taking a quick glance in the dataset?
1) It's a fixed width file - it will require a dictionary providing the width of each field;

2) It has three different kinds of table delimited by a "Format Column" - this will be hugely important while loading the algorithm and processing the data

3) We'll use Pandas to load data - it will require some adaptations to be computationally efficient;

### Step 1: Selecting what we want to load - use parsimony!
1. Select the variables that we want to open - do we really need to open the whole file?
2. Create a name for those selected variables;
3. Be carefull with long and heavy string variables;

In [5]:
#Width of the columns of table1:
colspecs = [(0,1),(3,17),(17,18),(367,375),
            (375,382),(682,684),(684,688),(688,738)]

In [6]:
#Name of the columns of table 1:
col_nam = ['TIPO DE REGISTRO','CNPJ', 'IDENTIFICADOR MATRIZ/FILIAL',
             'DATA INICIO ATIVIDADE','CEP', 'UF', 'CODIGO MUNICIPIO',
            'MUNICIPIO']

### Step 2: Creating the algorithm - efficiency is key over here
1. Initialize the algorithm based on its default values, dictionaries and also on the **chunksize** - this last item is very important to load data progressively allowing a fractionalization of the dataset;
2. Test the number of chunks on 0dataset - we've adopted about 1mn of rows for each subset of data, it will depends on your CPU and GPU power;
3. Put all data together based on a one-by-rest scheme - iterator of the algo - this will define the 'order' of your solution; 
4. Creating time-estimation variables - this would be interesting to preview the total time to handle with all data at once - your algorithm is trully efficient?

In [7]:
class table:
    def __init__(self, path):
        self.data = pd.read_fwf(path, colspecs, iterator = True, chunksize = 10**6, dtype='str', header = None)
            
    def test(self):
        for chunk in self.data:
            print(len(chunk))
            
    def table(self):
        self.df = pd.DataFrame()
        self.data = pd.read_fwf(path, colspecs, iterator = True, chunksize = 10**6, dtype='str', header = None)
        i = 0
        n = 0
        chunk_start = time.time()
        #test:
        for chunk in self.data:
            chunk.columns = col_nam
            chunk = chunk[(chunk["TIPO DE REGISTRO"]).astype(int) == 1]
            self.df = pd.concat([self.df, chunk])
            #Time elapsed:
            chunk_end = time.time()
            time_elapsed = chunk_end-chunk_start
            hours, rem = divmod(time_elapsed, 3600)
            minutes, seconds = divmod(rem, 60)
            #Total time (estimated):
            total_time = (time_elapsed/(i+1))*(72-i)
            hours2, rem2 = divmod(total_time, 3600)
            minutes2, seconds2 = divmod(rem2, 60)
            print('Chunk: ' + str(i) +" ; Time elapsed: " + "{:0>2}:{:0>2}:{:05.2f}".format(int(hours),int(minutes),seconds) +
                 " ; Total time (estimated): " + "{:0>2}:{:0>2}:{:05.2f}".format(int(hours2),int(minutes2),seconds2))
            #Counters
            i += 1
            n += 1
            #Testing
        end3 = time.time()
        hours3, rem3 = divmod(end3-chunk_start, 3600)
        minutes3, seconds3 = divmod(rem3, 60)
        print("Total time: " + "{:0>2}:{:0>2}:{:05.2f}".format(int(hours3),int(minutes3),seconds3))
        return self.df

In [8]:
path = 'E:/Dados Empresa/DADOS_ABERTOS_CNPJ/cnpj.txt'

In [11]:
#Initializing our algorithm with its default values:
tab = table(path)

In [12]:
#Creating the table in formats 1:
table = tab.table()

Chunk: 0 ; Time elapsed: 00:00:20.18 ; Total time (estimated): 00:24:13.07
Chunk: 1 ; Time elapsed: 00:01:00.58 ; Total time (estimated): 00:35:50.69
Chunk: 2 ; Time elapsed: 00:01:40.78 ; Total time (estimated): 00:39:11.43
Chunk: 3 ; Time elapsed: 00:02:21.38 ; Total time (estimated): 00:40:38.81
Chunk: 4 ; Time elapsed: 00:03:01.68 ; Total time (estimated): 00:41:10.91
Chunk: 5 ; Time elapsed: 00:03:42.45 ; Total time (estimated): 00:41:24.03
Chunk: 6 ; Time elapsed: 00:04:22.97 ; Total time (estimated): 00:41:19.41
Chunk: 7 ; Time elapsed: 00:05:03.89 ; Total time (estimated): 00:41:09.09
Chunk: 8 ; Time elapsed: 00:05:45.12 ; Total time (estimated): 00:40:54.18
Chunk: 9 ; Time elapsed: 00:06:26.03 ; Total time (estimated): 00:40:31.96
Chunk: 10 ; Time elapsed: 00:07:07.12 ; Total time (estimated): 00:40:07.39
Chunk: 11 ; Time elapsed: 00:07:48.11 ; Total time (estimated): 00:39:39.58
Chunk: 12 ; Time elapsed: 00:08:29.72 ; Total time (estimated): 00:39:12.57
Chunk: 13 ; Time elaps

### Step 3: Final Result - this is what you really need?

Now you've done all the dirty job to put all data together. However, it will require some extra manipulation to extract the relevant information for our analysis. This will be our aim for the next sections.

In [13]:
table

TIPO DE REGISTRO            CNPJ IDENTIFICADOR MATRIZ/FILIAL  \
1                       1  00000000000191                           1   
40                      1  00000000000272                           2   
41                      1  00000000000353                           2   
42                      1  00000000000434                           2   
43                      1  00000000000515                           2   
44                      1  00000000000604                           2   
45                      1  00000000000787                           2   
46                      1  00000000000868                           2   
47                      1  00000000000949                           2   
48                      1  00000000001082                           2   
49                      1  00000000001163                           2   
50                      1  00000000001244                           2   
51                      1  00000000001325                           2   
52                      1  00000000001406                           2   
53                      1  00000000001597                           2   
54                      1  00000000001678                           2   
55                      1  00000000001759                           2   
56                      1  00000000001830                           2   
57                      1  00000000001910                           2   
58                      1  00000000002054                           2   
59                      1  00000000002135                           2   
60                      1  00000000002216                           2   
61                      1  00000000002305                           2   
62                      1  00000000002488                           2   
63                      1  00000000002569                           2   
64                      1  00000000002640                           2   
65                      1  00000000002720                           2   
66                      1  00000000002801                           2   
67                      1  00000000002992                           2   
68                      1  00000000003026                           2   
...                   ...             ...                         ...   
72385878                1  98767908000120                           1   
72385879                1  98768179000126                           1   
72385880                1  98768187000172                           1   
72385881                1  98768195000119                           1   
72385882                1  98768211000173                           1   
72385883                1  98817851000126                           1   
72385884                1  98817992000149                           1   
72385885                1  98818198000110                           1   
72385886                1  98818214000174                           1   
72385889                1  98818271000153                           1   
72385890                1  98818370000135                           1   
72385891                1  98818487000119                           1   
72385892                1  98818529000111                           1   
72385893                1  98818834000103                           1   
72385894                1  98818883000146                           1   
72385895                1  98819006000190                           1   
72385896                1  98819121000164                           1   
72385897                1  98819188000107                           1   
72385898                1  98819295000127                           1   
72385899                1  98819303000135                           1   
72385900                1  98819402000117                           1   
72385903                1  98819402000206                           2   
72385904                1  98819410000163                          

## Section 5.2 - Manipulating your Dataset

### Section 5.2.1. Creating a summary table mapping the number of firms born over each county per year
Now, our objective is to create a summary table for our dataset. We're going to keep the most important concept in Data Mining: be objective! Keep only what you need - this would be easier while manipulating the data. 

#### Steps:
1. Keep only the variables that you're going to use on this table;
2. Clean your string variables;
3. Transform time and data format to your purpouses;
4. Develop a solution to pivot your sub-dataset;

In [18]:
df = table[['DATA INICIO ATIVIDADE', 'MUNICIPIO']]

In [24]:
# Step 1:
df = table[['DATA INICIO ATIVIDADE', 'MUNICIPIO']]

# Step 2:
df['MUNICIPIO'] = df['MUNICIPIO'].str.replace('\d+', '')

# Step 3:
def time_conversor(name, col, df):
    df[name] = pd.to_datetime(df[col], errors = 'coerce').dt.year
    df = df[pd.notnull(df['DATA'])]    
    df[name] = df[name].astype(int)
    return df
# Converting the time format:
df = time_conversor('DATA', 'DATA INICIO ATIVIDADE', df)

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """
C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':
C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas

In [25]:
# Step 4: manipulate your dataset using pandas' functionalities
class data_visualization:
    def __init__(self, df):
        self.df = df
    
    def frequency(self, ind, col, val):
        self.ind = ind
        self.col = col
        self.val = val
        self.freq = self.df.groupby([self.ind, self.col]).size().reset_index(name = self.val)
        return self.freq, self.ind, self.col, self. val
    
    def pivot(self, ind, col, val):
        self.freq, self.ind, self.col, self. val = self.frequency(ind, col, val)
        self.data = self.freq.reset_index()
        self.pivot = self.data.pivot_table(index = self.ind, columns = self.col, values = self.val, 
                                           aggfunc = 'sum', margins = True)
        self.pivot = self.pivot.fillna(0)
        return self.pivot

In [26]:
# Initializing algorithm
data = data_visualization(df)
# Creating the pivot-table:
database = data.pivot('DATA', 'MUNICIPIO', 'FREQ')

In [27]:
database

MUNICIPIO  ABADIA DE GOIAS  ABADIA DOS DOURADOS  ABADIANIA  ABAETE  \
DATA                                                                 
1696                   0.0                  0.0        0.0     0.0   
1891                   0.0                  0.0        0.0     0.0   
1893                   0.0                  0.0        0.0     0.0   
1899                   0.0                  0.0        0.0     0.0   
1901                   0.0                  0.0        0.0     0.0   
1902                   0.0                  0.0        0.0     0.0   
1903                   0.0                  0.0        0.0     0.0   
1904                   0.0                  0.0        0.0     0.0   
1905                   0.0                  0.0        0.0     0.0   
1906                   0.0                  0.0        0.0     0.0   
1907                   0.0                  0.0        0.0     0.0   
1908                   0.0                  0.0        0.0     0.0   
1909                   0.0                  0.0        0.0     0.0   
1910                   0.0                  0.0        0.0     0.0   
1911                   0.0                  0.0        0.0     0.0   
1912                   0.0                  0.0        0.0     0.0   
1913                   0.0                  0.0        0.0     0.0   
1914                   0.0                  0.0        0.0     0.0   
1915                   0.0                  0.0        0.0     0.0   
1916                   0.0                  0.0        0.0     0.0   
1917                   0.0                  0.0        0.0     0.0   
1918                   0.0                  0.0        0.0     0.0   
1919                   0.0                  0.0        0.0     0.0   
1920                   0.0                  0.0        0.0     0.0   
1921                   0.0                  0.0        0.0     0.0   
1922                   0.0                  0.0        0.0     0.0   
1923                   0.0                  0.0        0.0     0.0   
1924                   0.0                  0.0        0.0     0.0   
1925                   0.0                  0.0        0.0     0.0   
1926                   0.0                  0.0        0.0     0.0   
...                    ...                  ...        ...     ...   
1990                   9.0                 18.0       34.0   138.0   
1991                   5.0                 10.0       29.0   122.0   
1992                   8.0                 11.0       14.0    56.0   
1993                   5.0                 12.0       33.0    87.0   
1994                   2.0                 11.0       23.0    91.0   
1995                   6.0                 17.0       20.0   100.0   
1996                   6.0                 10.0       25.0   124.0   
1997                  28.0                 10.0       27.0   113.0   
1998                  17.0                 12.0       25.0    86.0   
1999                  40.0                  5.0       22.0    89.0   
2000                  32.0                  9.0       32.0    91.0   
2001                  35.0                 11.0       28.0    74.0   
2002                  30.0                  6.0       28.0    55.0   
2003                  36.0                  7.0       32.0    66.0   
2004                  76.0                 34.0       82.0   168.0   
2005                  41.0                 29.0       44.0    60.0   
2006                  38.0                 16.0       30.0    74.0   
2007                  30.0                 21.0       29.0    84.0   
2008                 122.0                 68.0       93.0   164.0   
2009                  47.0                 29.0       45.0   152.0   
2010                 105.0                 31.0       99.0   199.0   
2011                  87.0                 27.0      103.0   181.0   
2012                 214.0                110.0      180.0   300.0   
2013                 131.0                 50.0      164.0   233.0   
2014      

In [28]:
database.describe()

MUNICIPIO  ABADIA DE GOIAS  ABADIA DOS DOURADOS    ABADIANIA       ABAETE  \
count           123.000000           123.000000   123.000000   123.000000   
mean             34.422764            17.788618    40.065041    88.373984   
std             194.991005            99.467455   225.338315   491.433791   
min               0.000000             0.000000     0.000000     0.000000   
25%               0.000000             0.000000     0.000000     0.000000   
50%               0.000000             0.000000     0.000000     0.000000   
75%               4.000000            11.000000    20.500000    70.000000   
max            2117.000000          1094.000000  2464.000000  5435.000000   

MUNICIPIO   ABAETETUBA     ABAIARA      ABAIRA        ABARE       ABATIA  \
count       123.000000  123.000000  123.000000   123.000000   123.000000   
mean        162.211382    8.780488   12.292683    18.048780    20.617886   
std         909.346002   49.269672   68.580183   101.619013   115.031750   
min           0.000000    0.000000    0.000000     0.000000     0.000000   
25%           0.000000    0.000000    0.000000     0.000000     0.000000   
50%           0.000000    0.000000    0.000000     0.000000     0.000000   
75%          75.500000    4.000000    9.000000     5.000000    14.000000   
max        9976.000000  540.000000  756.000000  1110.000000  1268.000000   

MUNICIPIO  ABDON BATISTA      ...        XAVANTINA        XAXIM       XEXEU  \
count         123.000000      ...       123.000000   123.000000  123.000000   
mean            7.349593      ...        13.853659   103.642276   14.032520   
std            41.444409      ...        77.184756   576.333509   79.613506   
min             0.000000      ...         0.000000     0.000000    0.000000   
25%             0.000000      ...         0.000000     0.000000    0.000000   
50%             0.000000      ...         0.000000     0.000000    0.000000   
75%             2.000000      ...        11.500000    90.500000    1.000000   
max           452.000000      ...       852.000000  6374.000000  863.000000   

MUNICIPIO     XINGUARA  XIQUE-XIQUE      ZABELE     ZACARIAS      ZE DOCA  \
count       123.000000   123.000000  123.000000   123.000000   123.000000   
mean        103.479675    63.902439    4.146341    22.032520    56.585366   
std         579.006970   356.360138   23.776860   125.370267   316.301525   
min           0.000000     0.000000    0.000000     0.000000     0.000000   
25%           0.000000     0.000000    0.000000     0.000000     0.000000   
50%           0.000000     0.000000    0.000000     0.000000     0.000000   
75%          79.500000    50.500000    0.000000     2.000000    45.500000   
max        6364.000000  3930.000000  255.000000  1355.000000  3480.000000   

MUNICIPIO      ZORTEA           All  
count      123.000000  1.230000e+02  
mean        11.512195  6.380313e+05  
std         64.882032  3.557346e+06  
min          0.000000  1.000000e+00  
25%          0.000000  9.000000e+00  
50%          0.000000  2.000000e+02  
75%          1.000000  5.594385e+05  
max        708.000000  3.923892e+07  

[8 rows x 5290 columns]

### Do you notice something wrong on this dataset?
- The first firm established in Brazil was Santanense in 1891 - but why we have a firm in 1696?
- Why in 1901 we have a huge jump in terms of firms?

**Lesson 1: Real Data is not as perfect as we would wonder...**

In [69]:
#We're going to discard this strange observations
data = database.iloc[5:,:]

### Section 5.2.2: Adding some new features on the dataset

What if we add some geo-locational reference to the Brazilian counties?

As an example, I have merged CNPJ dataset with "Municipios do Brasil com Latitude e Longitude" available at [Qlik Brasil](https://community.qlik.com/t5/Qlik-Brasil/Municipios-do-Brasil-com-Latitude-e-Longitude/gpm-p/1480217). On the next notebook, I will plot this geo-referenced data on Brazil's map with the purpouse to introduce some data visualization libraries disposable in Python.

In [76]:
#Loading the auxiliar dataset
aux = pd.read_excel('d:/Users/leonardo.martins/Documents/Workshop - NEREUS/MunicipiosBrasil.xls')

In [77]:
# Initializing algorithm - the same as developed before
data2 = data_visualization(df_new)
# Creating the pivot-table:
data_new, _, _, _, = data2.frequency('DATA', 'MUNICIPIO', 'FREQ')

In [78]:
#Merging both datasets on the pre-processed data:
data_new = pd.merge(data_new, aux, how='left', left_on='MUNICIPIO',right_on='MUNICIPIO')

In [79]:
data_new

DATA               MUNICIPIO  FREQ  LONGITUDE  LATITUDE   UF
0       1696            JUIZ DE FORA     1     -43.35    -21.76   MG
1       1891           MONTES CLAROS     1     -43.86    -16.73   MG
2       1893          BELO HORIZONTE     1     -43.95    -19.81   MG
3       1899               SAO PAULO     1     -46.63    -23.54   SP
4       1901              ABRE CAMPO     1     -42.47    -20.30   MG
5       1901               ALAGOINHA     2     -35.54     -6.95   PB
6       1901               ALAGOINHA     2     -36.77     -8.46   PE
7       1901                ALEGRETE     1     -55.79    -29.78   RS
8       1901                ALMENARA     1     -40.69    -16.18   MG
9       1901                 ALTINHO     1     -36.05     -8.49   PE
10      1901                 ARACAJU     1     -37.07    -10.91   SE
11      1901                 ARATUBA     1     -39.04     -4.41   CE
12      1901          AURORA DO PARA     1     -47.56     -2.13   PA
13      1901                    BAGE     1     -54.10    -31.33   RS
14      1901                BANABUIU     1     -38.92     -5.31   CE
15      1901                   BELEM     6     -36.49     -9.57   AL
16      1901                   BELEM     6     -48.50     -1.45   PA
17      1901                   BELEM     6     -35.51     -6.74   PB
18      1901           BELMIRO BRAGA     1     -43.41    -21.94   MG
19      1901            BOM DESPACHO     1     -45.25    -19.73   MG
20      1901                BRASILIA    18     -47.93    -15.78   DF
21      1901  BREJO DA MADRE DE DEUS     1     -36.37     -8.14   PE
22      1901                 BREJOES     1     -39.79    -13.10   BA
23      1901                  BUIQUE     1     -37.15     -8.62   PE
24      1901         CACAPAVA DO SUL     1     -53.49    -30.51   RS
25      1901        CACHOEIRA DO SUL     1     -52.89    -30.03   RS
26      1901            CACHOEIRINHA     3     -36.23     -8.48   PE
27      1901            CACHOEIRINHA     3     -51.09    -29.95   RS
28      1901            CACHOEIRINHA     3     -47.92     -6.12   TO
29      1901                 CAMAQUA     2     -51.81    -30.85   RS
...      ...                     ...   ...        ...       ...  ...
268810  2018        VITORIA DO XINGU    58     -52.01     -2.88   PA
268811  2018                VITORINO   111     -52.78    -26.27   PR
268812  2018         VITORINO FREIRE    75     -45.23     -4.04   MA
268813  2018            VOLTA GRANDE    41     -42.53    -21.77   MG
268814  2018           VOLTA REDONDA  3760     -44.10    -22.52   RJ
268815  2018              VOTORANTIM  1765     -47.43    -23.54   SP
268816  2018             VOTUPORANGA  1437     -49.97    -20.42   SP
268817  2018                  WAGNER    23     -41.16    -12.28   BA
268818  2018             WALL FERRAZ     4     -41.91     -7.23   PI
268819  2018            WANDERLANDIA    57     -47.96     -6.84   TO
268820  2018               WANDERLEY    27     -43.88    -12.12   BA
268821  2018          WENCESLAU BRAZ   478     -45.36    -22.53   MG
268822  2018          WENCESLAU BRAZ   478     -49.80    -23.87   PR
268823  2018     WENCESLAU GUIMARAES    86     -39.47    -13.68   BA
268824  2018               WESTFALIA    33        NaN       NaN  NaN
268825  2018               WITMARSUM    51     -49.79    -26.92   SC
268826  2018                 XAMBIOA    87     -48.53     -6.41   TO
268827  2018                  XAMBRE    48     -53.49    -23.73   PR
268828  2018               XANGRI-LA   380     -50.04    -29.80   RS
268829  2018                 XANXERE   870     -52.40    -26.87   SC
268830  2018                  XAPURI    49     -68.50    -10.65   AC
268831  2018               XAVANTINA    19     -52.34    -27.06   SC
268832  2018                   XAXIM   289     -52.53    -26.96   SC
268833  2018                   XEXEU    50     -35.62     -8.80   PE
268834  2018                XINGUARA   322     -49.94     -7.09   PA
268835  2018             XIQUE-XIQUE   157     -42.73    -10.82   BA
268836

### Final Result: Data Tables

Now that we've both datasets, we're going to export them as xlsx to both save server space, and also allowing a more efficient computational processing.

In [75]:
data.to_excel('d:/Users/leonardo.martins/Documents/Workshop - NEREUS/dataset1.xlsx')
data_new.to_excel('d:/Users/leonardo.martins/Documents/Workshop - NEREUS/dataset2.xlsx')